### PR TITLE
rpcwallet: fix remote signing issues

### LINKED ISSUE #7022: wallet: check we can produce valid signature for address before sending funds on-chain
EDIT (guggero):

See [comment below](https://github.com/lightningnetwork/lnd/issues/7022#issuecomment-1278695682), this issue was turned from a bug report into a feature request.

Steps to implement:
 - Before sending any funds on chain, check each output pk script whether it belongs to the wallet.
 - For each output in a transaction that belongs to the wallet, produce a valid spend witness stack and validate it with a script VM to make sure the funds can be spent later on.

Original issue text:

> I have a RaspiBlitz running with the following specifications:
> 
> bitcoin: v23.0.0
> LND: 0.15.2-beta
> 
> A few days ago, I started using clightning in parallel to my LND node on my RaspiBlitz and wanted to now switch over the funds to that lightning node. 
> 
> I closed all channels that the node has, which worked as expected. Now my LND node reports on-chain funds as expected, and also respective UTXOs. However, when I try to sweep the on-chain funds to my normal bitcoin wallet, I get the following error:
> 
> 
> 
> > Unable To Broadcast Sweep Transaction: Unmatched Backend Error: -26: Non-mandatory-script-verify-flag (script Failed An Op_equalverify Operation)
> 
> 
> 
> I already did a rescan of the wallet, and it discovered the same UTXOs again. Can you please advise how I could regain access to my funds? Trying to open a channel to my new lightning instance (clightning) didn't work, it also fails to broadcast the funding transaction and creates channels that stay forever in 'pending' state.
> 
> Update: I managed to identify the UTXO which is causing a problem (by using the 'Lease' functionality in Ride the Lightning). Now there is only one UTXO left. However it has the largest balance so it would be critical to retrieve it.

### LINKED ISSUE #7024: [bug]: Unable to construct funding tx/derived public key does not match bip32 derivation info public key
### Background

We are using two lnd instances (remote signer feature) and trying to open a channel:

```
lncli openchannel --node_key 02015cb10a74914fe6c126e591b1a4bda66662ca807e97a94cf973ffe408259f74 --local_amt 3000000 --sat_per_vbyte 1
[lncli] rpc error: code = Unknown desc = unable to construct funding tx: error signing with remoteinstance: remote signer returned invalid key spend signature: signature too short with 0 bytes
```

The relevant logs on the watch only instance are:
```
2022-10-11 20:43:02.050 [INF] SWPR: Manual fee rate input of 250 sat/kw is too low, using 253 sat/kw instead
2022-10-11 20:43:02.050 [INF] FNDG: Initiating fundingRequest(local_amt=0.03 BTC (subtract_fees=false), push_amt=0 mSAT, chain_hash=000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f, peer=02015cb10a74914fe6c126e591b1a4bda66662ca807e97a94cf973ffe408259f74, min_confs=1)
2022-10-11 20:43:02.050 [WRN] LNWL: Web API does not have a fee rate for target=3, using the fee rate for target=2 instead
2022-10-11 20:43:02.050 [INF] CHFD: Performing funding tx coin selection using 253 sat/kw as fee rate
2022-10-11 20:43:02.065 [WRN] CHFD: Unable to find funding output for shim intent: unable to create witness script, no funding keys
2022-10-11 20:43:02.082 [INF] FNDG: Target commit tx sat/kw for pendingID(c2c839c49b0cc1ea2f1c8aa67c7e979c29f4cc3c219af6c48e50f9cf828f9699): 1275
2022-10-11 20:43:02.082 [INF] FNDG: Dust limit for pendingID(c2c839c49b0cc1ea2f1c8aa67c7e979c29f4cc3c219af6c48e50f9cf828f9699): 0.00000354 BTC
2022-10-11 20:43:02.082 [INF] FNDG: Starting funding workflow with lrttbpewpyf4w4ecprhgfih4ymhhssbnqmrcyjzbksjmwhxhc5ra63yd.onion:9736 for pending_id(c2c839c49b0cc1ea2f1c8aa67c7e979c29f4cc3c219af6c48e50f9cf828f9699), committype=tweakless
2022-10-11 20:43:03.311 [INF] FNDG: Recv'd fundingResponse for pending_id(c2c839c49b0cc1ea2f1c8aa67c7e979c29f4cc3c219af6c48e50f9cf828f9699)
2022-10-11 20:43:03.313 [ERR] FNDG: Unable to process contribution from &{{[2465652 16775426 43306903 33684062 40264394 6893417 40786203 20945668 17462417 22316]} {[59556220 3875047 1227694 23039521 940911 37078236 64646312 30028152 54437282 974180]}}: unable to construct funding tx: error signing with remoteinstance: remote signer returned invalid key spend signature: signature too short with 0 bytes
2022-10-11 20:43:03.313 [INF] FNDG: Cancelling funding reservation for node_key=02015cb10a74914fe6c126e591b1a4bda66662ca807e97a94cf973ffe408259f74, chan_id=c2c839c49b0cc1ea2f1c8aa67c7e979c29f4cc3c219af6c48e50f9cf828f9699
2022-10-11 20:43:03.313 [ERR] RPCS: unable to open channel to NodeKey(02015cb10a74914fe6c126e591b1a4bda66662ca807e97a94cf973ffe408259f74): unable to construct funding tx: error signing with remoteinstance: remote signer returned invalid key spend signature: signature too short with 0 bytes
2022-10-11 20:43:03.313 [ERR] RPCS: [/lnrpc.Lightning/OpenChannel]: unable to construct funding tx: error signing with remoteinstance: remote signer returned invalid key spend signature: signature too short with 0 bytes
```

If we look at the remote signer logs, the only relevant line is:
```
[WRN] BTWL: SignPsbt: Skipping input 0, derived public key 02464e51e1eede9721d0c97bc15cbfea5c38a6f6b820c69ac14623c971932464d3 does not match bip32 derivation info public key 02cd0efa576d888e03623a311e3192de0f2df06a88a0493ee64463c91954b3a075
```

The node is "identity_pubkey": "030e7d77f4012c809e4bcd0701e7d50b3a0869679f328e325e222ae56acabc74b2" which has dozens of channels. The only recent change was the update from 0.15.1 to 0.15.2 and some channels that are being force-closed.

### Environment
Both instances are using 0.15.2-beta
Backend is a full bitcoin core 22 in neutrino mode (`-txindex=1 -blockfilterindex=1 -peerblockfilters=1`)

### PR DESCRIPTION
**Pull request by guggero**
Fixes #7024.
Fixes #7022.

### CODE REVIEW

File: keychain/btcwallet.go [hunk 1]
```diff
@@ -473,7 +473,7 @@ func (b *BtcWalletKeyRing) SignMessageSchnorr(keyLoc KeyLocator,
 	}
 
 	if len(taprootTweak) > 0 {
-		privKey = txscript.TweakTaprootPrivKey(privKey, taprootTweak)
+		privKey = txscript.TweakTaprootPrivKey(*privKey, taprootTweak)
```
Comment (yyforyongyu) [inline keychain/btcwallet.go:476]: hmmm so the `privKey` can be mutated by others?

**Review by yyforyongyu — APPROVED** [review]
LGTM👍 Thanks for the quick fix! Just FYI the new itest failed due to flakes and is unrelated to this PR. The linter issue should be gone once the btcd is in. And maybe you could rebase on #7126 to see how the old itest goes?

File: keychain/btcwallet.go (hunk 1 above)
Comment (guggero) [reply keychain/btcwallet.go:476]: Yeah, tweaking ended up mutating the key. Schnorr signing too, so we'll want https://github.com/btcsuite/btcd/pull/1919 to also be included (which is why the itest would also currently fail, independent of #7126).

**Comment by guggero** [comment]
Pushed up an updated version that uses the latest dependencies. The itest won't be green before https://github.com/lightningnetwork/lnd/pull/7126 is merged but I rand the remote signing itests locally to confirm the targeted fixes are correct.

**Comment by guggero** [comment]
I tacked on a commit (29d86a51) that fixes another Taproot related issue in a remote signing setup (related to Pool account deposits).

**Comment by morehouse** [comment]
> Fixes #7022.

Maybe this fixes the original cause of #7022, but it seems we should leave that issue open to track work on double-checking change outputs.  Better safe than sorry, and it would keep users funds safe if we introduce a bug like this again.

**Comment by Roasbeef** [comment]
> Better safe than sorry, and it would keep users funds safe if we introduce a bug like this again.

Yeah there're def some subtle interactive here, it took quite a bit to get to the bottom of all this. 

@morehouse have you seen: https://github.com/btcsuite/btcwallet/pull/825? This PR updates the `btcwallet` dep that includes that fix. I have another idea re also storing the _serialized_ address, then decoding that before returning the addr to the caller to check if the checksum fails or not. Then `txscript.PayToAddrScript` can call this extra method or validate again there. I think this is about the best we can do in that situation.

**Comment by Roasbeef** [comment]
Ok merged the itest PR, can you rebase this now @guggero, then the new set of itests can run on top of it.

**Review by yyforyongyu — APPROVED** [review]

**Comment by guggero** [comment]
> > Fixes #7022.
> 
> Maybe this fixes the original cause of #7022, but it seems we should leave that issue open to track work on double-checking change outputs. Better safe than sorry, and it would keep users funds safe if we introduce a bug like this again.

@morehouse I'm not sure what you mean. This PR includes a bump in the btcwallet version that includes https://github.com/btcsuite/btcwallet/pull/825. Which is what you mean if I understand you correctly.

**Comment by morehouse** [comment]
> @morehouse have you seen: [btcsuite/btcwallet#825](https://github.com/btcsuite/btcwallet/pull/825)? This PR updates the `btcwallet` dep that includes that fix.

Ack, thanks for explaining.

> @morehouse I'm not sure what you mean. This PR includes a bump in the btcwallet version that includes https://github.com/btcsuite/btcwallet/pull/825. Which is what you mean if I understand you correctly.

Yep, sorry I missed the fix over there.

**Pull request merged**